### ISSUE TITLE
[BUG] Mod-taps frequently getting stuck

### ISSUE THREAD

**Issue by elkowar** [zmkfirmware/zmk#986]
I've noticed more and more that my home-row modifier mod-tap keys keep getting stuck from time to time. So I did some analysis.

Assuming both S and L are configured as mod-taps for LSHFT and RSHFT respectively, and D and K are both mod-taps for LGUI. A tapping term of 150 is being used.

the following inputs consistently result in the shift modifier being stuck (which is then only resolvable by hitting reset)
HOLD(s), WAIT(200ms), TAP(l), RELEASE(s)

additionally, the following inputs:
HOLD(s), WAIT(200ms), TAP(**k**), RELEASE(s)
result in no immediate change, until either d or k (either of the mod-taps for LGUI) are tapped again. As soon as either of them is tapped, the gui modifier starts being stuck as held down.

Looking at the logs, I observed the **error message** `ACTIVE_HOLD_TAP_CLEANED_UP_TOO_EARLY`

Here is some more log output (this is showing the first input pattern I describe):
<summary>
logs
<details>
<pre>
[00:00:51.001,953] <dbg> zmk.zmk_endpoints_send_report: usage page 0x07
[00:00:54.885,833] <dbg> zmk.kscan_matrix_read: Sending event at 1,2 state on
[00:00:54.885,894] <dbg> zmk.zmk_kscan_process_msgq: Row: 1, col: 2, position: 14, pressed: true
[00:00:54.885,925] <dbg> zmk.position_state_changed_listener: 14 bubble (no undecided hold_tap active)
[00:00:54.885,955] <dbg> zmk.position_state_down: combo: capturing position event 14
[00:00:54.885,955] <dbg> zmk.zmk_event_manager_handle_from: Listener captured the event
[00:00:56.227,569] <dbg> zmk.kscan_matrix_read: Sending event at 4,3 state on
[00:00:56.227,630] <dbg> zmk.zmk_kscan_process_msgq: Row: 4, col: 3, position: 21, pressed: true
[00:00:56.227,661] <dbg> zmk.position_state_changed_listener: 21 bubble (no undecided hold_tap active)
[00:00:56.227,691] <dbg> zmk.position_state_down: combo: capturing position event 21
[00:00:56.227,722] <dbg> zmk.release_pressed_keys: combo: releasing position event 14
[00:00:56.227,722] <dbg> zmk.zmk_keymap_apply_position_state: layer: 0 position: 14, binding name: MOD_TAP
[00:00:56.227,752] <dbg> zmk.on_hold_tap_binding_pressed: 14 new undecided hold_tap
[00:00:56.227,783] <dbg> zmk.release_pressed_keys: combo: reraising position event 21
[00:00:56.227,813] <dbg> zmk.decide_hold_tap: 14 decided hold-timer (tap-preferred decision moment timer)
[00:00:56.227,813] <dbg> zmk.on_keymap_binding_pressed: position 14 keycode 0x700e1
[00:00:56.227,874] <dbg> zmk.hid_listener_keycode_pressed: usage_page 0x07 keycode 0xe1 implicit_mods 0x00 explicit_mods 0x00
[00:00:56.227,874] <dbg> zmk.zmk_hid_register_mod: Modifier 1 count 1
[00:00:56.227,874] <dbg> zmk.zmk_hid_register_mod: Modifiers set to 0x02
[00:00:56.227,874] <dbg> zmk.zmk_hid_implicit_modifiers_press: Modifiers set to 0x02
[00:00:56.227,905] <dbg> zmk.zmk_endpoints_send_report: usage page 0x07
[00:00:56.227,966] <dbg> zmk.position_state_changed_listener: 536871936 capturing 21 down event
[00:00:56.227,966] <dbg> zmk.zmk_event_manager_handle_from: Listener captured the event
[00:00:56.227,996] <dbg> zmk.zmk_event_manager_handle_from: Listener captured the event
[00:00:56.351,837] <dbg> zmk.kscan_matrix_read: Sending event at 4,3 state off
[00:00:56.351,898] <dbg> zmk.zmk_kscan_process_msgq: Row: 4, col: 3, position: 21, pressed: false
[00:00:56.351,928] <dbg> zmk.position_state_changed_listener: 21 bubble (no undecided hold_tap active)
[00:00:56.351,959] <dbg> zmk.zmk_keymap_apply_position_state: layer: 0 position: 21, binding name: MOD_TAP
[00:00:56.351,989] <err> zmk: ACTIVE_HOLD_TAP_CLEANED_UP_TOO_EARLY
[00:00:56.796,051] <dbg> zmk.kscan_matrix_read: Sending event at 1,2 state off
[00:00:56.796,142] <dbg> zmk.zmk_kscan_process_msgq: Row: 1, col: 2, position: 14, pressed: false
[00:00:56.796,173] <dbg> zmk.position_state_changed_listener: 14 bubble (no undecided hold_tap active)
[00:00:56.796,203] <dbg> zmk.zmk_keymap_apply_position_state: layer: 0 position: 14, binding name: MOD_TAP
[00:00:56.796,234] <dbg> zmk.on_keymap_binding_released: position 14 keycode 0x700e1
[00:00:56.796,264] <dbg> zmk.hid_listener_keycode_released: usage_page 0x07 keycode 0xe1 implicit_mods 0x00 explicit_mods 0x00
[00:00:56.796,295] <dbg> zmk.zmk_hid_unregister_mod: Modifier 1 count: 0
[00:00:56.796,295] <dbg> zmk.zmk_hid_unregister_mod: Modifier 1 released
[00:00:56.796,295] <dbg> zmk.zmk_hid_unregister_mod: Modifiers set to 0x00
[00:00:56.796,295] <dbg> zmk.zmk_hid_implicit_modifiers_release: Modifiers set to 0x00
[00:00:56.796,325] <dbg> zmk.zmk_endpoints_send_report: usage page 0x07
[00:00:56.796,386] <dbg> zmk.on_hold_tap_binding_released: 14 cleaning up hold-tap
[00:00:58.210,021] <dbg> zmk.kscan_matrix_read: Sending event at 1,2 state on
[00:00:58.210,083] <dbg> zmk.zmk_kscan_process_msgq: Row: 1, col: 2, position: 14, pressed: true
[00:00:58.210,113] <dbg> zmk.position_state_changed_listener: 14 bubble (no undecided hold_tap active)
[00:00:58.210,144] <dbg> zmk.position_state_down: combo: capturing position event 14
[00:00:58.210,144] <dbg> zmk.zmk_event_manager_handle_from: Listener captured the event
[00:00:58.266,754] <dbg> zmk.kscan_matrix_read: Sending event at 1,2 state off
[00:00:58.266,876] <dbg> zmk.zmk_kscan_process_msgq: Row: 1, col: 2, position: 14, pressed: false
[00:00:58.266,906] <dbg> zmk.position_state_changed_listener: 14 bubble (no undecided hold_tap active)
[00:00:58.266,906] <dbg> zmk.release_pressed_keys: combo: releasing position event 14
[00:00:58.266,906] <dbg> zmk.zmk_keymap_apply_position_state: layer: 0 position: 14, binding name: MOD_TAP
[00:00:58.266,967] <dbg> zmk.on_hold_tap_binding_pressed: 14 new undecided hold_tap
[00:00:58.266,998] <dbg> zmk.zmk_keymap_apply_position_state: layer: 0 position: 14, binding name: MOD_TAP
[00:00:58.267,059] <dbg> zmk.decide_hold_tap: 14 decided tap (tap-preferred decision moment key-up)
[00:00:58.267,059] <dbg> zmk.on_keymap_binding_pressed: position 14 keycode 0x70016
[00:00:58.267,089] <dbg> zmk.hid_listener_keycode_pressed: usage_page 0x07 keycode 0x16 implicit_mods 0x00 explicit_mods 0x00
[00:00:58.267,120] <dbg> zmk.zmk_hid_implicit_modifiers_press: Modifiers set to 0x00
[00:00:58.267,120] <dbg> zmk.zmk_endpoints_send_report: usage page 0x07
[00:00:58.267,181] <dbg> zmk.release_captured_events: Releasing key position event for position 21 pressed
[00:00:58.267,181] <dbg> zmk.position_state_changed_listener: 21 bubble (no undecided hold_tap active)
[00:00:58.267,211] <dbg> zmk.position_state_down: combo: capturing position event 21
[00:00:58.267,211] <dbg> zmk.zmk_event_manager_handle_from: Listener captured the event
[00:00:58.267,242] <dbg> zmk.on_keymap_binding_released: position 14 keycode 0x70016
[00:00:58.267,272] <dbg> zmk.hid_listener_keycode_released: usage_page 0x07 keycode 0x16 implicit_mods 0x00 explicit_mods 0x00
[00:00:58.267,303] <dbg> zmk.zmk_hid_implicit_modifiers_release: Modifiers set to 0x00
[00:00:58.267,303] <dbg> zmk.zmk_endpoints_send_report: usage page 0x07
[00:00:58.275,054] <dbg> zmk.on_hold_tap_binding_released: 14 cleaning up hold-tap
[00:00:58.275,146] <dbg> zmk.release_pressed_keys: combo: releasing position event 21
[00:00:58.275,146] <dbg> zmk.zmk_keymap_apply_position_state: layer: 0 position: 21, binding name: MOD_TAP
[00:00:58.275,177] <dbg> zmk.on_hold_tap_binding_pressed: 21 new undecided hold_tap
[00:00:58.275,238] <dbg> zmk.decide_hold_tap: 21 decided hold-timer (tap-preferred decision moment timer)
[00:00:58.275,268] <dbg> zmk.on_keymap_binding_pressed: position 21 keycode 0x700e5
[00:00:58.275,329] <dbg> zmk.hid_listener_keycode_pressed: usage_page 0x07 keycode 0xe5 implicit_mods 0x00 explicit_mods 0x00
[00:00:58.275,329] <dbg> zmk.zmk_hid_register_mod: Modifier 5 count 1
[00:00:58.275,329] <dbg> zmk.zmk_hid_register_mod: Modifiers set to 0x20
[00:00:58.275,329] <dbg> zmk.zmk_hid_implicit_modifiers_press: Modifiers set to 0x20
[00:00:58.275,329] <dbg> zmk.zmk_endpoints_send_report: usage page 0x07
[00:00:58.603,515] <dbg> zmk.kscan_matrix_read: Sending event at 1,1 state on
[00:00:58.603,576] <dbg> zmk.zmk_kscan_process_msgq: Row: 1, col: 1, posi
</pre>
</details>
</summary>

**Comment by caksoylar** [comment]
Can you share your keymap that you reproduced this with? I tried to reproduce with a test case but haven't had luck so far. I am interested because I also get occasional stuck mods that I haven't been able to log but this sounds like a good lead.

**Comment by elkowar** [comment]
Sure, here's my config: https://github.com/elkowar/zmk-config/blob/master/config/boards/shields/choctopus44/choctopus44.keymap


@buffet had also reported stuck home row modifiers to me a few days before, but is not encountering them anymore - I assume my case is not the only way to run into this bug

**Comment by caksoylar** [comment]
Thanks, I don't see GUI mod-taps in that keymap but I guess they would be set up similar to the SHIFT ones.

**Comment by beta-tank** [comment]
I also got this bug on my [layout](https://github.com/beta-tank/zmk-config/blob/8a6d150a7a9c8b0fa1100f70fcd5d79b06a578bc/config/boards/shields/jiran/jiran.keymap): I have configured `&mt LSHFT SPACE` and `&mt RALT BSLH`. When I hold Shift and then tap Backslash my Alt gets stuck. I can only fix it by keyboard reboot..

**Comment by ad0ll** [comment]
This issue occurs about 20-40 times for me a day. My full keymaps are here:
- https://github.com/ad0ll/zmk-config/blob/main/kyria/kyria.keymap
- https://github.com/ad0ll/zmk-config/blob/main/lily58/lily58.keymap

I haven't done the same thorough investigation that elkowar has, but what I've observed is that I typically get stuck mods when I do a layer toggle.

**Comment by elkowar** [comment]
Yea, I still have the issue, and I've mostly had to pretty much just avoid using mod taps completely for now, which is kinda painful,...

**Comment by evantravers** [comment]
I _think_ I'm encountering this, I'm going to do some extra work to try and reproduce it reliably.

**Comment by okke-formsma** [comment]
@ad0ll @elkowar what hardware are you using?

**Comment by okke-formsma** [comment]
does this only happen if you have 'quick-release' and/or 'release-after-ms' configured?

**Comment by okke-formsma** [comment]
The accidental cleanup of the hold-tap must be happening in the timer (the other place where we clean up hold-taps has an additional line of debug prints. 

<img width="946" alt="image" src="https://user-images.githubusercontent.com/486179/152684725-a2df3808-d34c-4a16-ace5-2f43574dad06.png">

However, this code path can only be followed when `hold_tap->work_is_cancelled` and that is only set to `true` _after_ the behavior has been released.

<img width="942" alt="image" src="https://user-images.githubusercontent.com/486179/152684948-617e7341-1c24-4cca-9c4f-8c9d9c834f78.png">

**Comment by ad0ll** [comment]
@okke-formsma, I'm using Nice!Nano 2.0s and have tested with a Kyria, Corne, and Lily58. I have some unbuilt BlueMicros lying around that I could build, but I'd need a week-ish to get back to you.

My config for home row mods was this before I gave up:
`	hm: homerow_mods {            
			compatible = "zmk,behavior-hold-tap";            
			label = "HOMEROW_MODS";          
			#binding-cells = <2>;  
			tapping-term-ms = <129>;            
			quick_tap_ms = <400>;            
			flavor = "balanced";        
			bindings = <&kp>, <&kp>;        
		};    `
I had tried with retro-tap before, but I'm not seeing it in my config now.

I started hitting this issue about 70-100 times a day, and did a very naive pass at the code. I didn't get the same error messages at elkowar, but used their debugging efforts to make an attempt. The changes I made in [this commit ](https://github.com/ad0ll/zmk/commit/90dcd9e00e0ef252e390a247968ed1616d243512)(again, I don't know what I'm doing, but thought to experiment) reduced the occurrences down to about 5-8 a day. However, with these changes, when a mod tap issue was triggered, my entire keyboard would cease to work until a reset.

**Comment by okke-formsma** [comment]
After reading the logs in the initial post more carefully, I think the following sequence may be the root of the issue:

```
00:00:56.227,966]  zmk.position_state_changed_listener: 536871936 capturing 21 down event
[00:00:56.227,966]  zmk.zmk_event_manager_handle_from: Listener captured the event
[00:00:56.227,996]  zmk.zmk_event_manager_handle_from: Listener captured the event
[00:00:56.351,837]  zmk.kscan_matrix_read: Sending event at 4,3 state off
[00:00:56.351,898]  zmk.zmk_kscan_process_msgq: Row: 4, col: 3, position: 21, pressed: false
[00:00:56.351,928]  zmk.position_state_changed_listener: 21 bubble (no undecided hold_tap active)
[00:00:56.351,959]  zmk.zmk_keymap_apply_position_state: layer: 0 position: 21, binding name: MOD_TAP
[00:00:56.351,989]  zmk: ACTIVE_HOLD_TAP_CLEANED_UP_TOO_EARLY
```

So the '21 down' event is captured, but never released. The '21 up' event is processed too soon. Two seconds later, the keydown event for position 21 is released.

```
[00:00:58.267,181]  zmk.release_captured_events: Releasing key position event for position 21 pressed
```

This could be an interplay between combos and hold-taps 🤔

**Comment by okke-formsma** [comment]
This is weird:
<img width="681" alt="image" src="https://user-images.githubusercontent.com/486179/152702122-6df51c32-ba8e-4e68-9047-58b5bc71a0c9.png">

<img width="706" alt="image" src="https://user-images.githubusercontent.com/486179/152702134-a9f41c86-581b-4c68-993d-844bc7e96734.png">

So the position of the hold-tap that's supposedly capturing this keypress is `536871936`, which is not a valid position unless you have a _large_ board.

**Comment by evantravers** [comment]
That's a very clean number in binary: `536871936` -> `00100000000000000000010000000000`

Makes me wonder if there's an overflow, max int size, or similar curmudgeonry occurring.

**Comment by mrlinuxfish** [comment]
I'm getting this or a similar issue on my hummingbird. The strange thing is I haven't got this issue on any other board using a similar config

**Comment by okke-formsma** [comment]
@mrlinuxfish do you use different combos on your hummingbird?

**Comment by mrlinuxfish** [comment]
I have thumb combos (which are also on the Zaphod and and ble-Chiffre). The hummingbird also has overlapping combos on the bottom row to emulate the bottom corner keys

**Comment by mrlinuxfish** [comment]
I have not noticed any stuck keys after reducing my tapping term to 150 from 200. It may be coincidence or the timing window for this bug is smaller with the shorter tapping term. Either way the tighter tapping term works better for me overall even if it doesn't fix the issue

**Comment by elkowar** [comment]
> @ad0ll @elkowar what hardware are you using?

I'm using a choctopus44
I don't think I found any options that made a difference although I'm not 100% sure on those. Can't check right now, but I'll try to remember ^^'

Thanks for investigating!

**Comment by boekhocg** [comment]
Curious, anyone here experiencing this that was not connected to more than one profile? I connected to a second profile and started to constantly get the issue.  I disconnected from one and it's been good since. Might be coincidence, but thought I'd ask in case.

**Comment by trankillity** [comment]
> Curious, anyone here experiencing this that was not connected to more than one profile? I connected to a second profile and started to constantly get the issue. I disconnected from one and it's been good since. Might be coincidence, but thought I'd ask in case.

Unfortunately this isn't the case. I just tried to switch to Callum-style mods using sticky modifiers on layers on the home row and am experiencing stuck mods almost instantly upon flashing/using them. The weird thing is that I have a sticky shift on my base layer which has NEVER caused this issue.

I also have home row mods on my base layer which have never caused this issue either. Additionally, I have combos (and overlapping combos) against all of the keys used for home row mods and for the mods on the layers.

My full config [is here](https://github.com/trankillity/zmk-sweep) if it helps.

**Comment by halcyonCorsair** [comment]
I tried going back to pre-December 23 with no luck.

However I have had some luck by removing some combos:
With [this config](https://github.com/halcyonCorsair/zmk-config/commit/f9f9d0db67adbda1f7c70898001e50a5624ad121), I got the issue almost constantly.

However [these changes](https://github.com/halcyonCorsair/zmk-config/commit/98acf333ed7a5439af667b10e9dd59e33bd54dc3) seem to have made it go away.

**Comment by ad0ll** [comment]
Chiming in, I cut all of my combos and am still experiencing the issue. As for pre-December, I personally have had this issue consistently since getting nice nanos in Sept 2021

**Comment by trankillity** [comment]
Small anecdote to add to this issue:

I have just moved to using [Callum-style mods](https://github.com/callum-oakley/qmk_firmware/tree/master/users/callum) instead of mod-taps and the issue has completely disappeared. So definitely appears as if the issue is within the implementation of passing mods to the hold (or more likely, the release) portion of the `hold-tap` behaviour.

**Comment by timwenger** [comment]
@elkowar  and @okke-formsma.
I often get stuck down home row mods  (as in, until I reset, that modifier key is permanently being sent to the PC).
After some debugging, I can confirm that I have the same <err> tag showing ACTIVE_HOLD_TAP_CLEANED_UP_TOO_EARLY (log line 36), and I have the same corrupted key press number, 536871936 (log line 29), as the original poster.

I can reproduce this by typing in Notepad++ with the left hand (main side of the keyboard), which has 
	- a middle finger on index 15 [&hm LALT S] on Layer 0, and a simple [&kp DOWN] on layer 1.
	- a pinky finger on index 25 [&lt L1 Z] to go from layer 0 to layer 1.
	- a combo that includes a key at index 25, with a timeout-ms longer than the tapping-term-ms that is set for the corresponding &lt. my combo is: COMBO(num_one, 5000, 13 25, &kp N1), which sends num 1 when keys at index 13 and 25 are pressed within 5000 ms of each other. 5000ms exaggerated for exploiting this ticket.

To reproduce:
1. I press the pinky finger (index 25) to go into layer 1, then press the middle finger (index 15). Altogether, this should press the DOWN key.
2. However, if index 15 is pressed within 5000ms of index 25 being pressed, the DOWN key will not be pressed.

3. Now all keys can be released, it doesn't matter what order they are released in.

4. At this point, if index 15 is tapped, it prints one S, pauses for a quick moment, then prints S's continually until I hit another key, as if I tapped and then held index 15, the S key.
5. Regardless of whether you tapped index 15 in step 4, once any home row mod is tapped, the ALT key (the modifier on index 15) is now stuck (held down) until I reset the keyboard.
	After resetting, I have to hold index 15 (to produce an ALT keystroke) before the computer behaves normally again. Otherwise it behaves as though ALT is still pressed.
	
Therefore, I can affirm that a combo can affect homerow mod. This behavior was repeatable for other &hm keys, as well as repeatable for keys on the right half.

For now, my work around is to makes sure that my combo timeout-ms is not longer than my &lt tapping-term-ms.
I've attached my keymap file, and a debug log output of the events on steps 1 and 2.

[log.txt](https://github.com/zmkfirmware/zmk/files/8811296/log.txt)
[corne.keymap.txt](https://github.com/zmkfirmware/zmk/files/8811297/corne.keymap.txt)

I hope this helps debug this problem!

**Comment by urob** [comment]
Adding one more observation: This issue got significantly mitigated for me when using equal timeouts for all my combos. 

I used to have overlapping combos with different timeouts on almost all my alpha-keys, and my HRMs would get stuck virtually all the time. Setting all combo timeouts to the same value seemed to have at least mitigated the problem (so far, I haven't experienced any HRM getting stuck yet, but it's still early to tell). 

Possibly related: https://github.com/zmkfirmware/zmk/issues/905

**Comment by petejohanson** [comment]
Please feel free to re-open if you encounter this again w/ latest `main`. Thanks.

**Comment by w0uld** [comment]
Re-opening this. I've encountered this issue multiple times with different iterations of my configs, to the point where I'm no longer using the old one. 

Long story short: 
- Quick layer switches and mod-taps don't always work well together. I had accented keybinds on a secondary layer that I used fairly frequently, and when I used that layer with a shift/equal mod-tap, on release I sometimes had shift stuck until I pressed it again. I can't replicate it consistently, but it occurred every once in a while.
- I decided to switch to mod-morphs to see if the issue persisted. With mod-morphs instead of layer switches, the issue is non-existent. I stumbled across this bugreport only now by accident, but the issue seems very similar to what I've experienced, only with combos instead of layer switches. Let me know what I should share to help investigate this.
